### PR TITLE
net: relay I2P addresses even if not reachable (by us)

### PR DESCRIPTION
**Pull request by vasild**
Nodes that can reach the I2P network (have set `-i2psam=`) will relay
I2P addresses even without this patch. However, nodes that can't reach
the I2P network will not. This was done as a precaution in
https://github.com/bitcoin/bitcoin/pull/20119 before anybody could
connect to I2P because then, for sure, it would have been useless.

Now, however, we have I2P support and a bunch of I2P nodes, so get all
nodes on the network to relay I2P addresses to help with propagation,
similarly to what we do with Tor addresses.

### CODE REVIEW

**Comment by vasild** [comment]
cc @sipa

**Comment by vasild** [comment]
Would be good to get this in 22.0.

**Review by kristapsk — APPROVED** [review]
utACK ba45f0270815d54ae3290efc16324c2ff1984565

**Review by jonatack — COMMENTED** [review]
utACK ba45f0270815d54ae3290efc16324c2ff1984565 IsRelayable() is only called by RelayAddress() in the codebase. My I2P+onion+clearnet node has a slowly increasing number of I2P connections, both inbound and outbound, currently 9-10, but the number will increase once v22 is released and it seems good to relay them.

More test coverage may be good.

**Comment by vasild** [comment]
`ba45f02708..823b85e158`: append tests to ensure the behavior

> More test coverage may be good.

Here it is! :)

File: test/functional/p2p_addrv2_relay.py [hunk 1]
```diff
@@ -18,12 +18,19 @@
...
     addr = CAddress()
     addr.time = int(time.time()) + i
     addr.nServices = NODE_NETWORK | NODE_WITNESS
-    addr.ip = "123.123.123.{}".format(i % 256)
+    # Add one I2P address at an arbitrary position.
+    if i == 5:
+        addr.net = addr.NET_I2P
+        addr.ip = I2P_ADDR
+    else:
+        addr.ip = "123.123.123.{}".format(i % 256)
```
Comment (jonatack) [inline test/functional/p2p_addrv2_relay.py:33]: nit, f-strings for string interpolation

linter needs appeasing for "test/functional/p2p_addrv2_relay.py:19:1: F401 'test_framework.util.assert_equal' imported but unused"

Comment (jonatack) [reply test/functional/p2p_addrv2_relay.py:33]: will review in any case very shortly

File: test/functional/test_framework/messages.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 Classes use __slots__ to ensure extraneous attributes aren't accidentally added
 by tests, compromising their intended effect.
 """
+from base64 import b32encode, b32decode
```
Comment (jonatack) [inline test/functional/test_framework/messages.py:21]: nit, sort

File: test/functional/p2p_addrv2_relay.py [hunk 3]
```diff
@@ -64,9 +68,11 @@ def run_test(self):
...
         msg.addrs = ADDRS
         with self.nodes[0].assert_debug_log([
-                'Added 10 addresses from 127.0.0.1: 0 tried',
-                'received: addrv2 (131 bytes) peer=0',
-                'sending addrv2 (131 bytes) peer=1',
+                # The I2P address is not added to node's own addrman because it has no
+                # I2P reachability (thus 10 - 1 = 9).
+                'Added 9 addresses from 127.0.0.1: 0 tried',
+                'received: addrv2 (159 bytes) peer=0',
+                'sending addrv2 (159 bytes) peer=1',
```
Comment (jonatack) [inline test/functional/p2p_addrv2_relay.py:75]: (maybe) add an assert independent of the logging to back up the comment in lines 71-72, if you think it is useful
```diff
         assert addr_receiver.addrv2_received_and_checked
+        assert_equal(len(self.nodes[0].getnodeaddresses(count=0, network="i2p")), 0)
```

**Review by jonatack — COMMENTED** [review]
ACK 823b85e158161279f053e38935e7033d7d0a23a1 verified the test fails without ba45f0270815d54ae3290efc16324c2ff1984565 with `sending addrv2 (118 bytes) peer=1` (118 bytes instead of 159)

<details><summary>test failure without ba45f02</summary><p>

```rake
2021-06-11T16:15:19.507000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_znhw4kg4
2021-06-11T16:15:20.028000Z TestFramework (INFO): Create connection that sends addrv2 messages
2021-06-11T16:15:20.131000Z TestFramework (INFO): Send too-large addrv2 message
2021-06-11T16:15:20.188000Z TestFramework (INFO): Check that addrv2 message content is relayed and added to addrman
2021-06-11T16:15:22.319000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 128, in main
    self.run_test()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/p2p_addrv2_relay.py", line 78, in run_test
    addr_receiver.wait_for_addrv2()
  File "/usr/lib/python3.9/contextlib.py", line 124, in __exit__
    next(self.gen)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 400, in assert_debug_log
    self._raise_assertion_error('Expected messages "{}" does not partially match log:\n\n{}\n\n'.format(str(expected_msgs), print_log))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 166, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Expected messages "['Added 9 addresses from 127.0.0.1: 0 tried', 'received: addrv2 (159 bytes) peer=0', 'sending addrv2 (159 bytes) peer=1']" does not partially match log:

 - 2021-06-11T16:15:20.296366Z [msghand] [net_processing.cpp:2404] [ProcessMessage] received: addrv2 (159 bytes) peer=0
 - 2021-06-11T16:15:20.296802Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.0 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.296892Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.1 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.296976Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.2 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297026Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.3 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297075Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.4 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297122Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.6 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297165Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.7 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297217Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.8 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297266Z [msghand] [addrman.cpp:33] [GetNewBucket] IP 123.123.123.9 mapped to AS0 belongs to new bucket 628
 - 2021-06-11T16:15:20.297293Z [msghand] [addrman.h:532] [Add] Added 9 addresses from 127.0.0.1: 0 tried, 9 new
 - 2021-06-11T16:15:20.297604Z [msghand] [net_processing.cpp:2404] [ProcessMessage] received: ping (8 bytes) peer=0
 - 2021-06-11T16:15:20.297661Z [msghand] [net.cpp:2959] [PushMessage] sending pong (8 bytes) peer=0
 - 2021-06-11T16:15:20.347034Z [http] [httpserver.cpp:237] [http_request_cb] Received a POST request for / from 127.0.0.1:35120
 - 2021-06-11T16:15:20.347555Z [httpworker.0] [rpc/request.cpp:174] [parse] ThreadRPCServer method=setmocktime user=__cookie__
 - 2021-06-11T16:15:20.352683Z (mocktime: 2021-06-11T16:45:20Z) [opencon] [net.cpp:401] [ConnectNode] trying connection 123.123.123.8:8341 lastseen=2.5hrs
 - 2021-06-11T16:15:20.398215Z (mocktime: 2021-06-11T16:45:20Z) [msghand] [net.cpp:2959] [PushMessage] sending ping (8 bytes) peer=0
 - 2021-06-11T16:15:20.398608Z (mocktime: 2021-06-11T16:45:20Z) [msghand] [net.cpp:2959] [PushMessage] sending ping (8 bytes) peer=1
 - 2021-06-11T16:15:20.398852Z (mocktime: 2021-06-11T16:45:20Z) [msghand] [net.cpp:2959] [PushMessage] sending addrv2 (118 bytes) peer=1
 - 2021-06-11T16:15:20.399445Z (mocktime: 2021-06-11T16:45:20Z) [msghand] [net_processing.cpp:2404] [ProcessMessage] received: pong (8 bytes) peer=0
 - 2021-06-11T16:15:20.399591Z (mocktime: 2021-06-11T16:45:20Z) [msghand] [net_processing.cpp:2404] [ProcessMessage] received: pong (8 bytes) peer=1
```
</p></details>

**Comment by vasild** [comment]
`823b85e158...7593b06bd1`: address review suggestions and pet the linter

File: test/functional/p2p_addrv2_relay.py (hunk 3 above)
Comment (vasild) [reply test/functional/p2p_addrv2_relay.py:75]: Added.

File: test/functional/test_framework/messages.py (hunk 2 above)
Comment (vasild) [reply test/functional/test_framework/messages.py]: Done.

File: test/functional/p2p_addrv2_relay.py (hunk 1 above)
Comment (vasild) [reply test/functional/p2p_addrv2_relay.py]: Done.

File: test/functional/p2p_addrv2_relay.py (hunk 3 above)
Comment (jonatack) [reply test/functional/p2p_addrv2_relay.py:75]: Yes (in that case no need to pet the linter)

**Comment by jonatack** [comment]
ACK 7593b06bd1262f438bf34769ecc00e9c22328e56

**Review by kristapsk — APPROVED** [review]
ACK 7593b06bd1262f438bf34769ecc00e9c22328e56. Code looks correct, tested that functional test suite passes and also that `test/functional/p2p_addrv2_replay.py` fails if I undo changes in `IsRelayable()`.

**Comment by naumenkogs** [comment]
ACK 7593b06bd1262f438bf34769ecc00e9c22328e56.
Thank you for taking care of i2p.

**Comment by laanwj** [comment]
Code review ACK 7593b06bd1262f438bf34769ecc00e9c22328e56

**Pull request merged**